src/ReturnTypes/Helpers/AppExtension.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
+
+namespace NunoMaduro\Larastan\ReturnTypes\Helpers;
+
+use Illuminate\Foundation\Application;
+use NunoMaduro\Larastan\Concerns\HasContainer;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Expr\ClassConstFetch;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Name\FullyQualified;
+use PhpParser\Node\Scalar\String_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\ErrorType;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+use Throwable;
+
+class AppExtension implements DynamicFunctionReturnTypeExtension
+{
+    use HasContainer;
+
+    public function isFunctionSupported(FunctionReflection $functionReflection): bool
+    {
+        return $functionReflection->getName() === 'app';
+    }
+
+    public function getTypeFromFunctionCall(
+        FunctionReflection $functionReflection,
+        FuncCall $functionCall,
+        Scope $scope
+    ): Type {
+        if (count($functionCall->args) === 0) {
+            return new ObjectType(Application::class);
+        }
+
+        /** @var Expr $expr */
+        $expr = $functionCall->args[0]->value;
+
+        if ($expr instanceof String_) {
+            try {
+                return new ObjectType(get_class($this->resolve($expr->value)));
+            } catch (Throwable $exception) {
+                return new ErrorType();
+            }
+        }
+
+        if ($expr instanceof ClassConstFetch && $expr->class instanceof FullyQualified) {
```

Comment:
If we gonna resolve everything out of the container this will be unnecessary.

Or if you'd like, you can move this check above the `resolve` part. So if it's fully qualified class it will be returned directly, if not we will resolve.

I'm not sure which is better.

Reply:
If it is possible to simply resolve the $expr, then I'll get rid of this. (Unless there is a performance win by using the qualified name.)